### PR TITLE
TFTRT: Allow reshape from weights to constant scalar tensors

### PR DESCRIPTION
**Pull request by trevor-m**
TensorRT allows scalar tensors, which is denoted by setting nvinfer1::Dims to `nbDims=0, d=[empty]`. You are allowed to reshape a tensor with dims `nbDims=1, d=1` to a scalar.

TensorRT does not allow weights to be scalars, so a similar weight would have to have dims `nbDims=1, d=1`. But we can convert a weight into a tensor with IConstantLayer. Since we will then have a tensor, we can make it a scalar by setting `nbDims=0, d=[empty]`. However, we had a check in `PrepareTensorForShape` that prevented this. This change removes that unnecessary check.
 
This solves some conversion errors for object detection models.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -1297,7 +1297,10 @@ Status Converter::PrepareTensorForShape(const TRT_TensorOrWeights& input,
   // the dims are unknown or need to be inferred. And we don't do further checks
   // but rely on the caller to not make mistakes.
   // Otherwise we do simple check to make sure the total sizes are the same.
-  if (AreDimsStaticWithDifferentSize(input_dims, dims, input.is_tensor())) {
+  // If an input is a weight, it is going to become a tensor via
+  // CreateConstantLayer. So we can treat it as a tensor for
+  // AreDimsStaticWithDifferentSize(). This really only matters for 0-D tensors.
+  if (AreDimsStaticWithDifferentSize(input_dims, dims, /*is_tensor=*/true)) {
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Thanks for the fix, could you please add a unit test in the PrepareTensorForShape test suite to cover this use case?

Comment (trevor-m) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Yes!

**Comment by trevor-m** [comment]
@aaroey Could you please review the test cases? It would be great if we can get this merged before 1.14 is cut off.

**Review by aaroey — APPROVED** [review]

**Comment by aaroey** [comment]
@trevor-m sure, sorry for the delay.

**Pull request merged**